TypeCobol.Analysis/Cfg/ControlFlowGraphBuilder.cs

```diff
@@ -0,0 +1,2695 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using JetBrains.Annotations;
+using TypeCobol.Analysis.Graph;
+using TypeCobol.Compiler.CodeElements;
+using TypeCobol.Compiler.CodeModel;
+using TypeCobol.Compiler.CupParser.NodeBuilder;
+using TypeCobol.Compiler.Diagnostics;
+using TypeCobol.Compiler.Nodes;
+using TypeCobol.Compiler.Scopes;
+using TypeCobol.Compiler.Symbols;
+
+namespace TypeCobol.Analysis.Cfg
+{
+    /// <summary>
+    /// The Control Flow Graph Builder for a TypeCobol Program.
+    /// </summary>
+    public partial class ControlFlowGraphBuilder<D> : ProgramClassBuilderNodeListener
+    {
+        /// <summary>
+        /// CFG Modes.
+        /// Normal: is the normal mode, Paragraph blocs target by a PERFORM are not expanded.
+        /// Extended: Paragraph blocs target by a PERFORM are expanded.
+        /// </summary>
+        public enum CfgMode
+        {
+            Normal,
+            Extended,
+        }
+
+        /// <summary>
+        /// Set the CFG Mode
+        /// </summary>
+        public CfgMode Mode
+        {
+            get;
+            set;
+        }
+
+        /// <summary>
+        /// The parent program Control Flow Builder, for nested Program.
+        /// </summary>
+        public ControlFlowGraphBuilder<D> ParentProgramCfgBuilder
+        {
+            get;
+            private set;
+        }
+
+
+        /// <summary>
+        /// All Cfg graphs builder created during the building phase, so it contains Cfg for nested programs and nested procedures,
+        /// but also for stacked programs.
+        /// </summary>
+        public List<ControlFlowGraphBuilder<D>> AllCfgBuilder
+        {
+            get;
+            internal set;
+        }
+
+        /// <summary>
+        /// The current Program Cfg being built.
+        /// </summary>
+        public ControlFlowGraphBuilder<D> CurrentProgramCfgBuilder
+        {
+            get;
+            private set;
+        }
+
+        /// <summary>
+        /// The Control Flow Graph Build for the main Program
+        /// </summary>
+        public ControlFlowGraph<Node, D> Cfg
+        {
+            get;
+            private set;
+        }
+
+        /// <summary>
+        /// The Current Program symbol being built as a Scope
+        /// </summary>
+        public Program Program
```

Comment:
Duplicate with CurrentProgram proprerty, this can be safely removed.

Reply:
yes

Reply:
done